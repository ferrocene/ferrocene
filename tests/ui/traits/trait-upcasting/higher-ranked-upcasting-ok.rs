//@ revisions: current next
//@ ignore-compare-mode-next-solver (explicit revisions)
//@[next] compile-flags: -Znext-solver

// We should be able to instantiate a binder during trait upcasting.
// This test could be `check-pass`, but we should make sure that we
// do so in both trait solvers.
#![feature(trait_upcasting)]
#![crate_type = "rlib"]
trait Supertrait<'a, 'b> {}

trait Subtrait<'a, 'b>: Supertrait<'a, 'b> {}

impl<'a> Supertrait<'a, 'a> for () {}
impl<'a> Subtrait<'a, 'a> for () {}
fn ok(x: &dyn for<'a, 'b> Subtrait<'a, 'b>) -> &dyn for<'a> Supertrait<'a, 'a> {
    x //~ ERROR mismatched types
    //[current]~^ ERROR mismatched types
}
