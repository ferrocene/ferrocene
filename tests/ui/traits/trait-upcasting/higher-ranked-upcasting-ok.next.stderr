error[E0308]: mismatched types
  --> $DIR/higher-ranked-upcasting-ok.rs:17:5
   |
LL | fn ok(x: &dyn for<'a, 'b> Subtrait<'a, 'b>) -> &dyn for<'a> Supertrait<'a, 'a> {
   |                                                ------------------------------- expected `&dyn for<'a> Supertrait<'a, 'a>` because of return type
LL |     x
   |     ^ expected trait `Supertrait`, found trait `Subtrait`
   |
   = note: expected reference `&dyn for<'a> Supertrait<'a, 'a>`
              found reference `&dyn for<'a, 'b> Subtrait<'a, 'b>`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
